Sources/Leaf/LeafEncoder.swift

```diff
@@ -135,7 +135,7 @@ extension LeafEncoder {
     }
 
     private final class KeyedContainerImpl<Key>: KeyedEncodingContainerProtocol, LeafEncodingResolvable where Key: CodingKey {
-        private let encoder: EncoderImpl
+        private weak var encoder: EncoderImpl!
```

Comment:
We can't force unwrap there here - if they're weak then there's a chance that the reference will have been released when we try and use them, which will take down a server. So we'll need to check they exist where we use them

Reply:
except it was a let before, meaning it was guaranteed to be defined when used before the change.

Reply:
and actually yeah, the only initializer for these classes takes a non-optional encoder, so this should be safe:
`init(encoder: EncoderImpl) { self.encoder = encoder }`

Reply:
Except that the encoder is the only place where *ContainerImpl are created in the current codebase. Meaning that at the time of creation, and throughout its existence, its encoder must be defined. If the encoder is released, then yes it's released too, but that should be the only case where that happens, hence why all this does is break the cycle. Shouldn't be possible to crash the server in this case.